src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnDecoder.OctetString.cs

```diff
@@ -0,0 +1,673 @@
+        ///   otherwise, <see langword="false"/>.
+        /// </returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        ///   <paramref name="ruleSet"/> is not defined.
+        /// </exception>
+        /// <exception cref="AsnContentException">
+        ///   the next value does not have the correct tag.
+        ///
+        ///   -or-
+        ///
+        ///   the length encoding is not valid under the current encoding rules.
+        ///
+        ///   -or-
+        ///
+        ///   the contents are not valid under the current encoding rules.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        ///   <paramref name="expectedTag"/>.<see cref="Asn1Tag.TagClass"/> is
+        ///   <see cref="TagClass.Universal"/>, but
+        ///   <paramref name="expectedTag"/>.<see cref="Asn1Tag.TagValue"/> is not correct for
+        ///   the method.
+        /// </exception>
+        /// <seealso cref="TryReadOctetString"/>
+        public static bool TryReadPrimitiveOctetString(
+            ReadOnlySpan<byte> source,
+            AsnEncodingRules ruleSet,
+            out ReadOnlySpan<byte> value,
+            out int bytesConsumed,
+            Asn1Tag? expectedTag = null)
+        {
+            return TryReadPrimitiveOctetStringCore(
+                source,
+                ruleSet,
+                expectedTag ?? Asn1Tag.PrimitiveOctetString,
+                UniversalTagNumber.OctetString,
+                contentLength: out _,
+                headerLength: out _,
+                out value,
+                out bytesConsumed);
+        }
+
+        private static int CountConstructedOctetString(
+            ReadOnlySpan<byte> source,
+            AsnEncodingRules ruleSet,
+            bool isIndefinite)
+        {
+            int contentLength = CopyConstructedOctetString(
+                source,
+                ruleSet,
+                Span<byte>.Empty,
+                false,
+                isIndefinite,
+                out _);
+
+            // T-REC-X.690-201508 sec 9.2
+            if (ruleSet == AsnEncodingRules.CER && contentLength <= MaxCERSegmentSize)
+            {
+                throw new AsnContentException(SR.ContentException_InvalidUnderCerOrDer_TryBer);
+            }
+
+            return contentLength;
+        }
+
+        private static void CopyConstructedOctetString(
+            ReadOnlySpan<byte> source,
+            AsnEncodingRules ruleSet,
+            Span<byte> destination,
+            bool isIndefinite,
+            out int bytesRead,
+            out int bytesWritten)
+        {
+            bytesWritten = CopyConstructedOctetString(
+                source,
+                ruleSet,
+                destination,
+                true,
+                isIndefinite,
+                out bytesRead);
+        }
+
+        private static int CopyConstructedOctetString(
```

Comment:
This method might not behave correctly when _source_ and _destination_ point to the same address space. There are some public entry points (such as `TryReadPrimitiveOctetString` on line 255) that can lead to this condition.

How we want to handle this is up to you: detect and throw, or include logic to support it, or say "undefined behavior". But I think it'd be good to document that we've chosen one of these three worlds.